File: payments/laravel/app/Services/ProjectPayments.php
<?php

namespace App\Services;

use PDO;
use RuntimeException;

final class ProjectPayments
{
    public const CATALOG = [
        'monthly' => ['productKey'=>'pro','currency'=>'USD','unitAmount'=>1500,'type'=>'recurring','interval'=>'month'],
        'lifetime' => ['productKey'=>'pro','currency'=>'USD','unitAmount'=>9900,'type'=>'one_time','interval'=>null],
    ];

    private PDO $database;

    public function __construct(
        private readonly string $secret,
        private readonly string $environmentId,
        string $databasePath,
    ) {
        if (strlen($secret) < 32) throw new RuntimeException('Webhook secret must contain at least 32 bytes.');
        $directory = dirname($databasePath);
        if (!is_dir($directory)) mkdir($directory, 0770, true);
        $this->database = new PDO('sqlite:'.$databasePath, options: [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]);
        $this->database->exec('CREATE TABLE IF NOT EXISTS customer_subjects(customer_id TEXT PRIMARY KEY, subject TEXT NOT NULL UNIQUE); CREATE TABLE IF NOT EXISTS webhook_inbox(event_id TEXT PRIMARY KEY, body_digest TEXT NOT NULL); CREATE TABLE IF NOT EXISTS entitlements(subject TEXT NOT NULL, entitlement TEXT NOT NULL, quantity INTEGER NOT NULL, PRIMARY KEY(subject, entitlement));');
    }

    public function checkout(string $subject, array $input): array
    {
        if (array_keys($input) !== ['priceKey'] || !isset(self::CATALOG[$input['priceKey']])) throw new RuntimeException('Use one trusted catalog priceKey.');
        $customer = self::customerId($subject);
        $statement = $this->database->prepare('INSERT OR REPLACE INTO customer_subjects(customer_id, subject) VALUES (?, ?)');
        $statement->execute([$customer, $subject]);
        return ['url'=>sprintf('https://simulator.invalid/checkout/%s/%s/%s', rawurlencode($this->environmentId), rawurlencode($customer), rawurlencode($input['priceKey']))];
    }

    public function portal(string $subject): array
    {
        $customer = self::customerId($subject);
        $statement = $this->database->prepare('INSERT OR REPLACE INTO customer_subjects(customer_id, subject) VALUES (?, ?)');
        $statement->execute([$customer, $subject]);
        return ['url'=>sprintf('https://simulator.invalid/portal/%s/%s', rawurlencode($this->environmentId), rawurlencode($customer))];
    }

    public function acceptWebhook(string $rawBody, string $signature, ?int $now = null): bool
    {
        if (strlen($rawBody) > 131072) throw new RuntimeException('Webhook exceeds the size limit.');
        $parts = [];
        foreach (explode(';', $signature) as $part) { if (str_contains($part, '=')) { [$key,$value] = explode('=', $part, 2); $parts[$key]=$value; } }
        $timestamp = filter_var($parts['ts'] ?? null, FILTER_VALIDATE_INT);
        if ($timestamp === false || abs(($now ?? time()) - $timestamp) > 300) throw new RuntimeException('Invalid webhook signature.');
        $expected = hash_hmac('sha256', $timestamp.':'.$rawBody, $this->secret);
        if (!hash_equals($expected, $parts['h1'] ?? '')) throw new RuntimeException('Invalid webhook signature.');
        $event = json_decode($rawBody, true, flags: JSON_THROW_ON_ERROR);
        if (($event['data']['custom_data']['vibenest_environment_id'] ?? null) !== $this->environmentId) throw new RuntimeException('Webhook environment mismatch.');
        $this->validateEvent($event);
        $this->database->beginTransaction();
        try {
            $insert = $this->database->prepare('INSERT INTO webhook_inbox(event_id, body_digest) VALUES (?, ?)');
            $insert->execute([$event['event_id'], hash('sha256', $rawBody)]);
        } catch (\PDOException $error) {
            $this->database->rollBack();
            if ($error->getCode() === '23000') return false;
            throw $error;
        }
        if (($event['event_type'] ?? null) === 'transaction.completed') {
            $lookup = $this->database->prepare('SELECT subject FROM customer_subjects WHERE customer_id = ?');
            $lookup->execute([$event['data']['customer_id'] ?? null]);
            if ($subject = $lookup->fetchColumn()) {
                $grant = $this->database->prepare("INSERT INTO entitlements(subject, entitlement, quantity) VALUES (?, 'premium-access', 1) ON CONFLICT(subject, entitlement) DO UPDATE SET quantity=excluded.quantity");
                $grant->execute([$subject]);
            }
        }
        $this->database->commit();
        return true;
    }

    public function entitlementQuantity(string $subject): int
    {
        $query = $this->database->prepare("SELECT quantity FROM entitlements WHERE subject=? AND entitlement='premium-access'");
        $query->execute([$subject]);
        return (int) ($query->fetchColumn() ?: 0);
    }

    public static function customerId(string $subject): string { return 'sim_customer_'.substr(hash('sha256', $subject), 0, 24); }

    private function validateEvent(array $event): void
    {
        if (($event['event_type'] ?? null) !== 'transaction.completed') return;
        $data=$event['data'] ?? []; $items=$data['items'] ?? null;
        if (($data['status'] ?? null) !== 'completed' || !is_array($items) || count($items) !== 1) throw new RuntimeException('Invalid completed transaction.');
        $item=$items[0]; $price=self::CATALOG[$item['price_id'] ?? ''] ?? null; $totals=$data['details']['totals'] ?? [];
        if ($price===null || ($item['product_id'] ?? null)!==$price['productKey'] || ($item['quantity'] ?? null)!==1 || ($totals['total'] ?? null)!==(string)$price['unitAmount'] || ($totals['currency_code'] ?? null)!==$price['currency']) throw new RuntimeException('Transaction does not match the trusted catalog.');
    }
}

File: payments/rails/app/services/project_payments.rb
require "digest"
require "json"
require "openssl"
require "sqlite3"
require "uri"

class ProjectPayments
  CATALOG = {
    "monthly" => { productKey: "pro", currency: "USD", unitAmount: 1500, type: "recurring", interval: "month" },
    "lifetime" => { productKey: "pro", currency: "USD", unitAmount: 9900, type: "one_time", interval: nil }
  }.freeze

  def initialize(secret:, environment_id:, database_path:)
    raise ArgumentError, "Webhook secret must contain at least 32 bytes" if secret.bytesize < 32
    @secret = secret; @environment_id = environment_id
    FileUtils.mkdir_p(File.dirname(database_path))
    @database = SQLite3::Database.new(database_path)
    @database.execute_batch("CREATE TABLE IF NOT EXISTS customer_subjects(customer_id TEXT PRIMARY KEY, subject TEXT NOT NULL UNIQUE); CREATE TABLE IF NOT EXISTS webhook_inbox(event_id TEXT PRIMARY KEY, body_digest TEXT NOT NULL); CREATE TABLE IF NOT EXISTS entitlements(subject TEXT NOT NULL, entitlement TEXT NOT NULL, quantity INTEGER NOT NULL, PRIMARY KEY(subject, entitlement));")
  end

  def checkout(subject, input)
    raise ArgumentError, "Use one trusted catalog priceKey" unless input.keys == ["priceKey"] && CATALOG.key?(input["priceKey"])
    customer = self.class.customer_id(subject); bind(customer, subject)
    { url: "https://simulator.invalid/checkout/#{escape(@environment_id)}/#{escape(customer)}/#{escape(input['priceKey'])}" }
  end

  def portal(subject)
    customer = self.class.customer_id(subject); bind(customer, subject)
    { url: "https://simulator.invalid/portal/#{escape(@environment_id)}/#{escape(customer)}" }
  end

  def accept_webhook(raw_body, signature, now: Time.now.to_i)
    raise ArgumentError, "Webhook exceeds the size limit" if raw_body.bytesize > 128 * 1024
    parts = signature.split(";").filter_map { |part| part.split("=", 2) if part.include?("=") }.to_h
    timestamp = Integer(parts.fetch("ts")) rescue nil
    expected = timestamp && OpenSSL::HMAC.hexdigest("SHA256", @secret, "#{timestamp}:#{raw_body}")
    unless timestamp && (now - timestamp).abs <= 300 && secure_equal(expected, parts["h1"])
      raise ArgumentError, "Invalid webhook signature"
    end
    event = JSON.parse(raw_body)
    raise ArgumentError, "Webhook environment mismatch" unless event.dig("data", "custom_data", "vibenest_environment_id") == @environment_id
    validate_event(event)
    @database.execute("BEGIN IMMEDIATE")
    begin
      @database.execute("INSERT INTO webhook_inbox(event_id, body_digest) VALUES (?, ?)", [event.fetch("event_id"), Digest::SHA256.hexdigest(raw_body)])
      if event["event_type"] == "transaction.completed"
        subject = @database.get_first_value("SELECT subject FROM customer_subjects WHERE customer_id=?", [event.dig("data", "customer_id")])
        @database.execute("INSERT INTO entitlements(subject, entitlement, quantity) VALUES (?, 'premium-access', 1) ON CONFLICT(subject, entitlement) DO UPDATE SET quantity=excluded.quantity", [subject]) if subject
      end
      @database.execute("COMMIT")
    rescue SQLite3::ConstraintException
      @database.execute("ROLLBACK")
      return false
    rescue StandardError
      @database.execute("ROLLBACK")
      raise
    end
    true
  end

  def entitlement_quantity(subject)
    @database.get_first_value("SELECT quantity FROM entitlements WHERE subject=? AND entitlement='premium-access'", [subject]).to_i
  end

  def close = @database.close

  def self.customer_id(subject) = "sim_customer_" + Digest::SHA256.hexdigest(subject)[0, 24]
  private
  def bind(customer, subject) = @database.execute("INSERT OR REPLACE INTO customer_subjects(customer_id, subject) VALUES (?, ?)", [customer, subject])
  def escape(value) = URI.encode_www_form_component(value).gsub("+", "%20")
  def secure_equal(left, right) = left.is_a?(String) && right.is_a?(String) && left.bytesize == right.bytesize && ActiveSupport::SecurityUtils.secure_compare(left, right)
  def validate_event(event)
    return unless event["event_type"] == "transaction.completed"
    data = event["data"] || {}; items = data["items"]
    raise ArgumentError, "Invalid completed transaction" unless data["status"] == "completed" && items.is_a?(Array) && items.length == 1
    item = items.first; price = CATALOG[item["price_id"]]; totals = data.dig("details", "totals") || {}
    unless price && item["product_id"] == price[:productKey] && item["quantity"] == 1 && totals["total"] == price[:unitAmount].to_s && totals["currency_code"] == price[:currency]
      raise ArgumentError, "Transaction does not match the trusted catalog"
    end
  end
end
